javaslang/src/main/java/javaslang/collection/Iterator.java

```diff
@@ -467,51 +470,79 @@ public Short getNext() {
-        } else if (step * (from - toExclusive) >= 0) {
+        } else if (areEqual(from, toExclusive) || step.signum() == from.subtract(toExclusive).signum()) {
             return empty();
         } else {
-            return new AbstractIterator<Double>() {
+            if (step.compareTo(BigDecimal.ZERO) > 0) {
+                return new AbstractIterator<Double>() {
+                    BigDecimal i = from;
 
-                double prev = Double.NaN;
-                double curr = from;
-                boolean hasNext = true;
+                    @Override
+                    public boolean hasNext() {
+                        return i.compareTo(toExclusive) < 0;
+                    }
 
-                @Override
-                public boolean hasNext() {
-                    return hasNext;
-                }
+                    @Override
+                    public Double getNext() {
+                        final BigDecimal next = this.i;
+                        this.i = next.add(step);
+                        return next.doubleValue();
+                    }
+                };
+            } else {
+                return new AbstractIterator<Double>() {
+                    BigDecimal i = from;
+
+                    @Override
+                    public boolean hasNext() {
+                        return i.compareTo(toExclusive) > 0;
+                    }
 
-                @Override
-                public Double getNext() {
-                    final double next = curr;
-                    if ((step > 0 && curr + step >= toExclusive) || (step < 0 && curr + step <= toExclusive)) {
-                        hasNext = false;
-                    } else {
-                        prev = curr;
-                        curr += step;
-                        if (curr == prev) {
-                            hasNext = false;
-                        }
+                    @Override
+                    public Double getNext() {
+                        final BigDecimal next = this.i;
+                        this.i = next.add(step);
+                        return next.doubleValue();
                     }
-                    return next;
-                }
-            };
+                };
+            }
+        }
+    }
+
+    // TODO these two need a place to crash for the night :/
+
+    /* scale-independent equality */
+    static boolean areEqual(BigDecimal from, BigDecimal toExclusive) {
+        return from.compareTo(toExclusive) == 0;
+    }
+
+    /* parse infinite values also */
+    static final BigDecimal infinityMultiplier = BigDecimal.valueOf(1.000000000000001); // decrease by as little as possible so that it doesn't have a double representation anymore -> infinite
+
+    static BigDecimal asDecimal(double number) {
+        if (number == NEGATIVE_INFINITY) {
+            final BigDecimal result = BigDecimal.valueOf(Math.nextUp(NEGATIVE_INFINITY));
+            return result.multiply(infinityMultiplier);
```

Comment:
Yeah, may tests are failing that way, e.g. `shouldCreateRangeClosedByWhereFromIsLessThanToAndStepCorrectDirection`